sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala

```diff
@@ -85,24 +91,23 @@ case class DropFunction(
   extends RunnableCommand {
 
   override def run(sqlContext: SQLContext): Seq[Row] = {
+    val catalog = sqlContext.sessionState.catalog
     if (isTemp) {
       if (databaseName.isDefined) {
         throw new AnalysisException(
           s"It is not allowed to provide database name when dropping a temporary function. " +
             s"However, database name ${databaseName.get} is provided.")
       }
-      sqlContext.sessionState.catalog.dropTempFunction(functionName, ifExists)
+      catalog.dropTempFunction(functionName, ifExists)
     } else {
-      // We are dropping a permanent.
-      val dbName = databaseName.getOrElse(sqlContext.sessionState.catalog.getCurrentDatabase)
+      // We are dropping a permanent function.
+      val dbName = databaseName.getOrElse(catalog.getCurrentDatabase)
       val func = FunctionIdentifier(functionName, Some(dbName))
-      if (!ifExists) {
-        if (!sqlContext.sessionState.catalog.functionExists(func)) {
-          throw new AnalysisException(
-            s"Function '$functionName' does not exist in database '$dbName'.")
-        }
+      if (!ifExists && !catalog.functionExists(func)) {
+        throw new AnalysisException(
+          s"Function '$functionName' does not exist in database '$dbName'.")
```

Comment:
I still think this should be handled within `dropFunction` itself (by passing the `ifExists` flag), but not a big deal